internal/backend/runtime/omni/controllers/omni/redacted_cluster_machine_config.go

```diff
@@ -25,59 +31,279 @@ import (
+	}
+}
+
+// Name implements controller.QController interface.
+func (ctrl *RedactedClusterMachineConfigController) Name() string {
+	return "RedactedClusterMachineConfigController"
+}
 
+// Settings implements controller.QController interface.
+func (ctrl *RedactedClusterMachineConfigController) Settings() controller.QSettings {
+	return controller.QSettings{
+		RunHook: func(ctx context.Context, logger *zap.Logger, r controller.QRuntime) error {
+			ticker := ctrl.options.Clock.NewTicker(ctrl.options.DiffCleanupInterval)
+			defer ticker.Stop()
+
+			for {
+				select {
+				case <-ctx.Done():
 					return nil
+				case <-ticker.Chan():
+					if err := ctrl.cleanup(ctx, r, logger); err != nil {
+						return err
+					}
 				}
+			}
+		},
+		Inputs: []controller.Input{
+			{
+				Namespace: resources.DefaultNamespace,
+				Type:      omni.ClusterMachineConfigType,
+				Kind:      controller.InputQPrimary,
+			},
+		},
+		Outputs: []controller.Output{
+			{
+				Type: omni.RedactedClusterMachineConfigType,
+				Kind: controller.OutputExclusive,
+			},
+			{
+				Type: omni.MachineConfigDiffType,
+				Kind: controller.OutputExclusive,
+			},
+		},
+	}
+}
 
-				config, err := configloader.NewFromBytes(data)
-				if err != nil {
-					return err
-				}
+// Reconcile implements controller.QController interface.
+func (ctrl *RedactedClusterMachineConfigController) Reconcile(ctx context.Context, logger *zap.Logger, r controller.QRuntime, ptr resource.Pointer) error {
+	cmc, err := safe.ReaderGet[*omni.ClusterMachineConfig](ctx, r, ptr)
+	if err != nil {
+		if state.IsNotFoundError(err) {
+			return nil
+		}
 
-				redactedData, err := config.RedactSecrets(x509.Redacted).EncodeBytes(encoder.WithComments(encoder.CommentsDisabled))
-				if err != nil {
-					return err
-				}
+		return err
+	}
 
-				if err = cmcr.TypedSpec().Value.SetUncompressedData(redactedData); err != nil {
-					return err
-				}
+	if cmc.Metadata().Phase() == resource.PhaseTearingDown {
+		return ctrl.reconcileTearingDown(ctx, r, cmc)
+	}
 
-				helpers.CopyAllLabels(cmc, cmcr)
+	return ctrl.reconcileRunning(ctx, r, logger, cmc)
+}
 
-				return nil
-			},
-		},
-	)
+// MapInput implements controller.QController interface.
+func (ctrl *RedactedClusterMachineConfigController) MapInput(_ context.Context, _ *zap.Logger, _ controller.QRuntime, pointer resource.Pointer) ([]resource.Pointer, error) {
+	switch pointer.Type() {
+	case omni.ClusterMachineConfigType:
+		return []resource.Pointer{
+			omni.NewRedactedClusterMachineConfig(resources.DefaultNamespace, pointer.ID()).Metadata(),
+		}, nil
+	default:
+		return nil, fmt.Errorf("unexpected input type %q", pointer.Type())
+	}
+}
+
+func (ctrl *RedactedClusterMachineConfigController) reconcileRunning(ctx context.Context, r controller.ReaderWriter, logger *zap.Logger, cmc *omni.ClusterMachineConfig) error {
+	if err := r.AddFinalizer(ctx, cmc.Metadata(), ctrl.Name()); err != nil {
```

Comment:
It's better to check that there's no finalizer with the name on the resource, as AddFinalizer does extra Get call